connectors/sources/salesforce.py

```diff
@@ -0,0 +1,1475 @@
+        query_builder.with_fields(queryable_fields)
+        query = query_builder.build()
+
+        async for records in self._yield_non_bulk_query_pages(query):
+            for record in records:
+                sobjects[record["Id"]] = record
+
+        return sobjects
+
+    async def _is_queryable(self, sobject):
+        """User settings can cause sobjects to be non-queryable
+        Querying these causes errors, so we try to filter those out in advance
+        """
+        return sobject.lower() in await self.queryable_sobjects()
+
+    async def _select_queryable_fields(self, sobject, fields):
+        """User settings can cause fields to be non-queryable
+        Querying these causes errors, so we try to filter those out in advance
+        """
+        sobject_fields = await self.queryable_sobject_fields()
+        queryable_fields = sobject_fields.get(sobject, [])
+        return [f for f in fields if f.lower() in queryable_fields]
+
+    def _parse_content_documents(self, record):
+        content_docs = []
+        content_links = record.get("ContentDocumentLinks", {}) or {}
+        content_links = content_links.get("records", []) or []
+        for content_link in content_links:
+            content_doc = content_link.get("ContentDocument")
+            if not content_doc:
+                continue
+
+            content_doc["linked_sobject_id"] = record.get("Id")
+            content_docs.append(content_doc)
+
+        return content_docs
+
+    def _combine_duplicate_content_docs(self, content_docs):
+        """Duplicate ContentDocuments may appear linked to multiple SObjects
+        Here we ensure that we don't download any duplicates while retaining links"""
+        grouped = {}
+
+        for content_doc in content_docs:
+            content_doc_id = content_doc["Id"]
+            if content_doc_id in grouped:
+                linked_id = content_doc["linked_sobject_id"]
+                linked_ids = grouped[content_doc_id]["linked_ids"]
+                if linked_id not in linked_ids:
+                    linked_ids.append(linked_id)
+            else:
+                grouped[content_doc_id] = content_doc
+                grouped[content_doc_id]["linked_ids"] = [
+                    content_doc["linked_sobject_id"]
+                ]
+                # the id is now in the list of linked_ids so we can delete it
+                del grouped[content_doc_id]["linked_sobject_id"]
+
+        grouped_objects = list(grouped.values())
+        return grouped_objects
+
+    async def _yield_non_bulk_query_pages(self, soql_query):
+        """loops through query response pages and yields lists of records"""
+        url = f"{self.base_url}{QUERY_ENDPOINT}"
+        params = {"q": soql_query}
+
+        while True:
+            response = await self._get_json(
+                url,
+                params=params,
+            )
+            yield response.get("records")
+            if response.get("done", True) is True:
```

Comment:
Why defaults to `True`? Will there be scenario where `done` is not returned?

Reply:
I'm just being defensive in my coding. I'm pretty sure there will always be a `done` field.

Reply:
My point is, should it be defaulted to `False`?

Reply:
If it's defaulted to `False` it will continue trying to get the next page, which might not exist and will raise an error. But maybe that is better, if their API got weird we probably don't want to silently ingest half their dataset.

Alternative: keep it as `True` but log a warning if the `done` field is missing. WDYT?

Reply:
Or we could ignore it, and just rely on `nextRecordsUrl`.

We stop if there's no `nextRecordsUrl`.

Reply:
That's a lot simpler, I'll do that